### PR TITLE
Core benchmarks

### PR DESCRIPTION
**Pull request by xThaid**
As I said in the meeting, some of the benchmarks are disabled for now. All of the benchmarks have a magic constant that tells how many times a particular benchmark must run. This is to normalize running time. However some of the benchmarks runs only once, as they perform heavy computations. The benchmarks were created with an assumption that they will run at frequency at least 1MHz. Since verilator is able to simulate the core at only around 5KHz, I had to adjust these constants (by setting the parameter `cpu_mhz`) so we don't have to wait hours to finish. As a consequence, we can't run some of the benchmarks, because we can't go below one iteration.

todo: The toolchain must be updated for the benchmarks to compile.

### CODE REVIEW

**Review by tilk — APPROVED** [review]
Seems to look fine, and fits nicely into the testing workflow.

**Review by piotro888 — APPROVED** [review]
Looks great!

 Additional note: I think I found a small bug during review in https://github.com/kuznia-rdzeni/coreblocks/blob/481d8e0dc327e37eb9eea6ce656a0e273d2ab4eb/test/regression/memory.py#L116 (should be `SegmentFlags.WRITE`?)
It is indirectly related to this PR, so may be fixed here on in another PR.

File: .github/workflows/benchmark.yml [hunk 1]
```diff
@@ -7,7 +7,8 @@ on:
   workflow_dispatch:
 
 jobs:
-  build:
+  synthesis:
+    name: Synthesis senchmarks
```
Comment (tilk) [inline .github/workflows/benchmark.yml:11]:
```suggestion
    name: Synthesis benchmarks
```

**Comment by tilk** [comment]
I'm not merging, because the workflow failed when run manually.

https://github.com/kuznia-rdzeni/coreblocks/actions/runs/5120847462/jobs/9207866075

**Comment by lekcyjna123** [comment]
Still failing. This time there was a timeout. How long it takes on your PC to run tests? From my observations on GitHub it takes 4-6 times longer to execute the same scope.

File: test/external/embench/board_config/coreblocks-sim/boardsupport.h
[the remark this answers is not in the document]
Comment (xThaid) [reply test/external/embench/board_config/coreblocks-sim/boardsupport.h:1]: I thought it is needed by the embench, but apparently it is not.

File: test/external/embench/board_config/coreblocks-sim/boardsupport.c [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+static uint64_t cycle_cnt_start;
+static uint64_t instr_cnt_start;
+
+void start_trigger() {
+    cycle_cnt_start = rdcycle();
+    instr_cnt_start = rdinstret();
+}
+
+void stop_trigger() {
+    TO_HOST.cycle_cnt = rdcycle() - cycle_cnt_start;
```
[the remark this answers is not in the document]
Comment (xThaid) [reply test/external/embench/board_config/coreblocks-sim/boardsupport.c:28]: It doesn't really matter. The benchmark is like 500k-1000k cycles and subtraction probably <100.

File: test/external/embench/board_config/coreblocks-sim/crt0.S [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  li  x31,0
+
+.option push
+.option norelax
+  la gp, __global_pointer$
+.option pop
+
+.init_bss:
+  la a0, __bss_start
+  la a1, __bss_end-4 // section end is actually the start of the next section
```
[the remark this answers is not in the document]
Comment (xThaid) [reply test/external/embench/board_config/coreblocks-sim/crt0.S:45]: Is it not? Linker script is the place where you define alignment of all program sections. And C extension wouldn't matter here anyway. It is related to instruction memory.

File: test/regression/benchmark.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+    def write(self, req: WriteRequest) -> WriteReply:
+        if req.addr == 0x0:
+            self.on_finish()
+            return WriteReply()
+        else:
+            return super().write(req)
+
+    def return_code(self):
+        return int.from_bytes(self.data[4:8], "little")
```
[the remark this answers is not in the document]
Comment (xThaid) [reply test/regression/benchmark.py:26]: It is not related to the linker script, but rather to the C code that accesses these parts of memories. I'll add a clarifying comment.

**Comment by xThaid** [comment]
I excluded more benchmarks and decreased the constant, so benchmarks should run even faster. We can re-enable the benchmarks if we have a better machine.

**Comment by piotro888** [comment]
```
- name: Set ownership (Github Actions workaround)
        run: |
          # https://github.com/actions/runner/issues/2033
          chown -R $(id -u):$(id -g) $PWD
```
this step probably needs to be added to `run-perf-benchmarks` as well, to fix the commit issue.

**Pull request merged**